yb-voyager/src/queryparser/ddl_parser.go

```diff
@@ -0,0 +1,861 @@
+		time_range1 tsrange
+		EXCLUDE USING gist (room_id WITH =, time_range WITH &&),
+		EXCLUDE USING gist (room_id1 WITH =, time_range1 WITH &&)
+	);
+
+create_stmt:{relation:{relname:"Test" inh:true relpersistence:"p" location:14} table_elts:...table_elts:{constraint:{contype:CONSTR_EXCLUSION
+location:226 exclusions:{list:{items:{index_elem:{name:"room_id" ordering:SORTBY_DEFAULT nulls_ordering:SORTBY_NULLS_DEFAULT}}
+items:{list:{items:{string:{sval:"="}}}}}} exclusions:{list:{items:{index_elem:{name:"time_range" ordering:SORTBY_DEFAULT nulls_ordering:SORTBY_NULLS_DEFAULT}}
+items:{list:{items:{string:{sval:"&&"}}}}}} access_method:"gist"}} table_elts:{constraint:{contype:CONSTR_EXCLUSION location:282 exclusions:{list:
+{items:{index_elem:{name:"room_id1" ordering:SORTBY_DEFAULT nulls_ordering:SORTBY_NULLS_DEFAULT}} items:{list:{items:{string:{sval:"="}}}}}}
+exclusions:{list:{items:{index_elem:{name:"time_range1" ordering:SORTBY_DEFAULT nulls_ordering:SORTBY_NULLS_DEFAULT}} items:{list:{items:{string:{sval:"&&"}}}}}}
+access_method:"gist"}} oncommit:ONCOMMIT_NOOP}} stmt_len:365}
+
+here we are iterating over all the table_elts - table elements and which are comma separated column info in
+the DDL so each column has column_def(column definition) in the parse tree but in case it is a constraint, the column_def
+is nil.
+
+e.g. In case if PRIMARY KEY is included in column definition
+
+	 CREATE TABLE example2 (
+	 	id numeric NOT NULL PRIMARY KEY,
+		country_code varchar(3),
+		record_type varchar(5)
+
+) PARTITION BY RANGE (country_code, record_type) ;
+stmts:{stmt:{create_stmt:{relation:{relname:"example2"  inh:true  relpersistence:"p"  location:193}  table_elts:{column_def:{colname:"id"
+type_name:{names:{string:{sval:"pg_catalog"}}  names:{string:{sval:"numeric"}}  typemod:-1  location:208}  is_local:true
+constraints:{constraint:{contype:CONSTR_NOTNULL  location:216}}  constraints:{constraint:{contype:CONSTR_PRIMARY  location:225}}
+location:205}}  ...  partspec:{strategy:PARTITION_STRATEGY_RANGE
+part_params:{partition_elem:{name:"country_code"  location:310}}  part_params:{partition_elem:{name:"record_type"  location:324}}
+location:290}  oncommit:ONCOMMIT_NOOP}}  stmt_location:178  stmt_len:159}
+
+In case if PRIMARY KEY in column list CREATE TABLE example1 (..., PRIMARY KEY(id,country_code) ) PARTITION BY RANGE (country_code, record_type);
+stmts:{stmt:{create_stmt:{relation:{relname:"example1"  inh:true  relpersistence:"p"  location:15}  table_elts:{column_def:{colname:"id"
+type_name:{names:{string:{sval:"pg_catalog"}}  names:{string:{sval:"numeric"}}  ... table_elts:{constraint:{contype:CONSTR_PRIMARY
+location:98  keys:{string:{sval:"id"}} keys:{string:{sval:"country_code"}}}}  partspec:{strategy:PARTITION_STRATEGY_RANGE
+part_params:{partition_elem:{name:"country_code" location:150}}  part_params:{partition_elem:{name:"record_type"  ...
+*/
+func (p *TableParser) Parse(parseTree *pg_query.ParseResult) (DDLObject, error) {
```

Comment:
General thumb rule is: you wouldn't want to create a struct _just_ because you want to group a bunch of functions together. Only if the operations you do on the struct maintain some state over time, then it makes sense to create a struct. 

IMO, all these can be simple functions  `ParseTable` , `ParseForeignTable` , etc. 
Even the ddl detectors tbh, can be simple functions, but you want them to follow a certain function signature, so it's okay to create an interface, and a struct that follows that function signature. 

Similar to my other comment, do we really need `DDLParser`? If the detectors directly do the parsing, would we still need this?

Reply:
Agree to Aneesh's point. Having simple functions makes the code simpler.
But interfaces does help in enforcing the functions signatures, and provide some generalisation in code.

I was also thinking to suggest to avoid interface for DDLParser, maybe keep it for DDLIssueDetector.
See if that is easily possible to simplify that, otherwise interface is also fine.

Reply:
My Idea was to have the `DDLParser` interface and not simple functions for each ddl type is that we can utilize the DDLParser interface in other areas of code for example `export-schema/ import-schema` where if we want to parse a particular ddl type and get the object with information to modify it and also add some `Deparse()` function to get modified DDL. 
So with that, I feel we can keep the logic of figuring out the type of parser based on node type in `queryparser` only